rtt_rosparam/include/rtt_rosparam/RosParamDataSource.hpp

```diff
@@ -0,0 +1,221 @@
+  //   if (muse_cached_) {
+  //     if (!ros::param::getCached(mparam_name_, return_value)) {
+  //       throw(std::range_error("The parameter " + mparam_name_ + " cannot be fetched."));
+  //       return std::string();
+  //     }
+  //   } else {
+  //     if (!ros::param::get(mparam_name_, return_value)) {
+  //       throw(std::range_error("The parameter " + mparam_name_ + " cannot be fetched."));
+  //       return std::string();
+  //     }
+  //   }
+  //   base::DataSourceBase::shared_ptr mobj = this->clone();
+  //   if (nullptr == mobj) {
+  //     std::cerr << "Couldn't cast the ds ptr into a base::DataSourceBase::shared_ptr" << std::endl;
+  //   }
+  //   // From DataSource.cpp:48
+  //   // std::ostream mystream("stream");
+  //   // mobj->getTypeInfo()->write(std::cout, mobj);
+  //   std::cout << "DataSource typed: " << mobj->getTypeInfo()->getTypeIdName() << std::endl;
+  //   return return_value;
+  // }
+  
+  // template<>
+  // internal::DataSource<double>::result_t RosParamDataSource<double>::get() const
+  // {
+  //   double return_value;
+  //   std::cout << " It is a double ! " << std::endl;
+  //   if (muse_cached_) {
+  //     if (!ros::param::getCached(mparam_name_, return_value)) {
+  //       throw(std::range_error("The parameter " + mparam_name_ + " cannot be fetched."));
+  //       return 0.0;
+  //     }
+  //   } else {
+  //     if (!ros::param::get(mparam_name_, return_value)) {
+  //       throw(std::range_error("The parameter " + mparam_name_ + " cannot be fetched."));
+  //       return 0.0;
+  //     }
+  //   }
+  //   return return_value;
+  // }
+
+  template<typename T>
+  RosParamDataSource<T>::~RosParamDataSource() {}
+
+  template<typename T>
+  RosParamDataSource<T>::RosParamDataSource( std::string param_name, bool use_cached)
+      : mparam_name_(param_name),
+        muse_cached_(use_cached),
+        mcached_data_(T())
+  {
+  }
+  template<typename T>
+  void RosParamDataSource<T>::set( typename internal::AssignableDataSource<T>::param_t t )
+  {
+    ros::param::set(mparam_name_, t);
+  }
+
+  // template <typename T>
+  // typename internal::AssignableDataSource<T>::reference_t RosParamDataSource<T>::set() {
+  //   // return mdummy_;
+  // }
+
+  template<typename T>
+  RosParamDataSource<T>* RosParamDataSource<T>::clone() const
+  {
+      return new RosParamDataSource<T>(this->mparam_name_);
+  }
+
+  template<typename T>
+  RosParamDataSource<T>* RosParamDataSource<T>::copy( std::map<const base::DataSourceBase*, base::DataSourceBase*>& alreadyCloned ) const {
+      return const_cast<RosParamDataSource<T>*>(this); // no copy needed, data is outside.
+  }
+
+  } // namespace rosparam
+} //namespace RTT
+
+// /*
+//  * Extern template declarations for core data source types
+//  * (instantiated in DataSources.cpp)
+// */
+RTT_EXT_IMPL template class RTT::rosparam::RosParamDataSource< bool >;
+RTT_EXT_IMPL template class RTT::rosparam::RosParamDataSource< std::string >;
+
```

Comment:
```suggestion
```

This snippet obviously has been copied from [rtt/internal/DataSource.hpp](https://github.com/orocos-toolchain/rtt/blob/600102e8be9c81905b20930e32d43b28244ab173/rtt/internal/DataSource.hpp#L240). It is not needed here. In case of `RTT::internal::DataSource<T>` those data sources are already used internally in within RTT, which is why they are declared `extern` for every user of RTT. But that is not the case here.

However, you could use extern template declarations for all supported ROS parameter types here and explicitly instantiate them in the source, but then the comment is wrong and the list of types is incomplete. Easier to define the template class as-is and let the compiler return an error if it is used with a template type argument which is not supported by `ros::param::get()` or `ros::param::set()`.